### ISSUE TITLE
Review/refactor DataSet normalization methods

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/nd4j#889]
The DataSet provides two normalization methods.
https://github.com/deeplearning4j/nd4j/blob/8a0f25eca491fa190dc1df4caa02f8a4d570c4a2/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/dataset/DataSet.java#L549

These methods are:
- normalize()
- normalizeZeroMeanZeroUnitVariance()

In both cases, the normalization is conducted on each DataSet (minibatch) separately/independently. Any time the user has more than one minibatch (i.e., train/test, or data that can't fit in memory, or whatever) these methods are no longer useful.

Specifically, there's a few problems with this approach:
- It's not possible to get the means or stdevs from the training data to apply to the test data
- It's not useful for data that doesn't fit into memory all at once.
- I've observed that these methods can be confusing for users (newer users especially) - it's too easy to do the wrong type of normalization (on each minibatch separately), and too hard to do the *right* thing.

Suggestions:
1. Deprecate these normalization methods
2. Review the best method for data normalization (my thinking: ensure we have a proper set of DataSetPreProcessors (min-max, zero mean unit variance, etc), that can be built from a DataSetIterator (i.e., means, stdevs etc over entire training set), serialized and re-loaded)
3. Add a dedicated documentation page on DL4J website for normalization

**EDIT**: This has the sort of functionality I wask talking about: https://github.com/deeplearning4j/nd4j/blob/cf9fe733ace449bc797fb6b25879d8e5aa1683f1/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/dataset/api/iterator/StandardScaler.java
Just need to make that a DataSetPreProcessor and add some other options.

**EDIT 2**: Re: masking for RNN data, we can't just do mean/stdev along dimensions... we need to not include the time steps that are masked out, when calculating means, standard deviations, mins/maxes, etc.

**Comment by eraly** [comment]
I am going to add that we are going to have a way to make the normalization take a dimension or dimensions. Different sorts of datasets will require normalization along different dimensions. Eg - Images will require last two dimensions. But for time series it will be 0,2.

**Comment by eraly** [comment]
Closing issue. Masking/support for time series pending. But tracked in another issue.
